robotsim.py

```diff
@@ -48,118 +48,198 @@
-            y2 = y1 - int(math.sin(math.radians(angle)) * i)
-            x2 = 0 if x2 < 0 else display_width if x2 > display_width else x2
-            y2 = 0 if y2 < 0 else display_height if y2 > display_height else y2
-            is_wall = gameDisplay.get_at((x2,y2)) == colors['black']
-            pygame.draw.line(gameDisplay, red, (x1, y1), (x2, y2))
-            if is_wall:
-                pygame.display.update()
-                clock.tick(120)
-                return i
-        return self.sensor_range
-
-    def ultrasonic_forward(self):
-        return self.get_distance(self.w)
-
-    def ultrasonic_left(self):
-        return self.get_distance(self.w + 90)
-
-    def ultrasonic_back(self):
-        return self.get_distance(self.w + 180)
-
-    def ultrasonic_right(self):
-        return self.get_distance(self.w + 270)
-
-    def get_color(self):
-        x = self.x + int(pixel_constant * 0.5)
-        y = self.y + int(pixel_constant * 0.5)
-        generate_map()
-        c = gameDisplay.get_at((x,y))
-        self.set_position(self.x,self.y,self.w)
-        for name, value in colors.items():
-            if value == c:
-                return name
+        if 0 <= self.dir <= 2:
+            self.dir += 1
+        else:
+            self.dir = 0
+        for _ in range(30):
+            generate_map()
+            self.set_position(self.x,self.y,self.w + 3)
+
+    def ultrasonicFront(self):
+        distance = None
+        start = 0
+        if self.dir == 0:
+            # row-- until 0
+            for pos in range(self.row, -1, -1):
+                if map.tiles[pos][self.col].North.status in [1, 2]:
+                    distance = start
+                    break
+                start += 1
+            if distance == None:
+                distance = self.row
+            
+        if self.dir == 1:
+            # col-- until 0 
+            for pos in range(self.col, -1, -1):
+                if map.tiles[self.row][pos].West.status in [1, 2]:
+                    distance = start
+                    break
+                start += 1
+            if distance == None:
+                distance = self.col
+
+        if self.dir == 2:
+            # row++ until max
+            for pos in range(self.row, map.height):
+                if map.tiles[pos][self.col].South.status in [1, 2]:
+                    distance = start
+                    break
+                start += 1
+            if distance == None:
+                distance = map.height - self.row - 1
+
+        if self.dir == 3:
+            # col++ until 0
+            for pos in range(self.col, map.width):
+                if map.tiles[self.row][pos].East.status in [1, 2]:
+                    distance = start
+                    break
+                start += 1
+            if distance == None:
+                distance = map.width - self.col - 1 
+        pygame.display.update()
+        clock.tick(120)
+        return distance
```

Comment:
Great Implementation! I was just thinking that maybe it would be better to multiply it by 30 before returning it, just to clarify that this function returns you the cm to the next front wall/door.

Reply:
Thanks! I agree.